calyx-py/calyx/queues.py

```diff
@@ -149,37 +142,311 @@ def __len__(self) -> int:
         return self.pifo_len
 
 
-def operate_queue(commands, values, queue, max_cmds, keepgoing=False):
-    """Given the two lists, one of commands and one of values.
-    Feed these into our queue, and return the answer memory.
+@dataclass
+class Pieo:
+    """A PIEO data structure.
+    Supports the operations `push`, `pop`, and `peek`.
+
+    Stores elements ordered increasingly by a totally ordered `rank` attribute (for
+    simplicitly, our implementation is just using integers).
+
+    At initialization we take in a set of `(int, int, int)` triples `data` which stores
+    values, ready times, and their ranks, and is ordered by rank.
+    
+    We also take at initialization a `max_len` value to store the maximum possible
+    length of a queue.
+
+    When asked to push:
+    - If the PIEO is at length `max_len`, we raise an error.
+    - Otherwise, we insert the element into the PIEO such that the rank order stays increasing.
+
+    When asked to pop:
+    - If the length of `data` is 0, we raise an error .
+
+    - We can either pop based on value or based on eligibility.
+    - This implementation supports the most common eligibility predicate - whether an element is 'ripe'.
+
+    - If a value is passed in, we pop the first (lowest-rank) instance of that value.
+    - If no value is passed in but a time is,
+        we pop the first (lowest-rank) value that passes the predicate.
+    - Note that either a value or a bound must be passed in - both cannot be, nor can neither.
+
+    When asked to peek:
+    We do the same thing as `pop`, except:
+    - We peek into the PIEO instead of popping it - i.e. we don't remove any elements.
+
+    We compactly represent these similar operations through `query`, which takes in an additional
+    optional `remove` parameter (defaulted to False) to determine whether to pop or peek.
```

Comment:
This docstring is coming along nicely, but it needs more love. I think you need at first an explanation of what a PIEO is, generally. You can link to a website if you wish. Then please explain that we are not supporting full-on predicates, but rather a comparison against some time.

Also, "ripe" is a word I made up! If you wish to use it, define it first!